File: Silver/2016/US Open/03_closing_the_farm/ClosingTheFarm2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.*;

/**
 * @author Red
 * @version 1.0
 */
public class ClosingTheFarm2 {
    static int n;
    static int m;
    static List<List<Integer>> graph = new ArrayList<>();
    static boolean[] visited;
    static boolean[] closed;
    static Map<Integer, List<Integer>> regionIdToFarm = new HashMap<>();
    static Map<Integer, Integer> farmToRegionId = new HashMap<>();

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("closing.in"));
        PrintWriter pw = new PrintWriter("closing.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(r.readLine());
            int a = Integer.parseInt(st.nextToken()) - 1;
            int b = Integer.parseInt(st.nextToken()) - 1;
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        visited = new boolean[n];
        // 遍历图记录连通图数量
        int regionId = 0;
        for (int i = 0; i <n;i++){
            if (!visited[i]){
                regionIdToFarm.put(regionId, new ArrayList<>());
                dfs(i, regionId);
                regionId++;
            }
        }
        if (regionId == 1){
            pw.println("YES");
        } else {
            pw.println("NO");
        }
        // 依次关闭 barn
        // 遍历 barn 所在连通图中的其他 barn 构成新的连通图数量
        closed = new boolean[n];
        for (int i = 0; i < n - 1; i++) {
            int closedFarm = Integer.parseInt(r.readLine()) - 1;
            int farmRegionId = farmToRegionId.get(closedFarm);
            List<Integer> farmsInSameRegion = regionIdToFarm.get(farmRegionId);
            closed[closedFarm] = true;
            visited = closed.clone();
            regionIdToFarm.remove(farmRegionId);
            for (int farm : farmsInSameRegion){
                if (!visited[farm]){
                    regionIdToFarm.put(regionId, new ArrayList<>());
                    dfs(farm, regionId);
                    regionId++;
                }
            }
            if (regionIdToFarm.size() == 1){
                pw.println("YES");
            } else {
                pw.println("NO");
            }
        }
        pw.close();
    }

    /**
     * 深度优先搜索遍历图 并记录连通图id
     */
    static void dfs(int current, int regionId){
        if (visited[current]){
            return;
        }
        visited[current] = true;
        farmToRegionId.put(current, regionId);
        regionIdToFarm.get(regionId).add(current);
        for (int next : graph.get(current)){
            if (!visited[next]){
                dfs(next, regionId);
            }
        }

    }
}

File: Silver/2016/US Open/03_closing_the_farm/closing_the_farm2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

int n, m;
vector<vector<int>> graph;
vector<bool> visited;
vector<bool> closed;
map<int, vector<int>> region_id_to_farm;
map<int, int> farm_to_region_id;

void dfs(int current, int region_id);

int main() {
    ifstream fin("closing.in");
    ofstream fout("closing.out");

    fin >> n >> m;

    graph.resize(n);
    visited.resize(n, false);
    closed.resize(n, false);

    for (int i = 0; i < m; i++) {
        int a, b;
        fin >> a >> b;
        a--, b--; // Adjusting to 0-based indexing
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    int region_id = 0;

    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            region_id_to_farm[region_id] = vector<int>();
            dfs(i, region_id);
            region_id++;
        }
    }

    if (region_id == 1) {
        fout << "YES" << endl;
    } else {
        fout << "NO" << endl;
    }

    for (int i = 0; i < n - 1; i++) {
        int closed_farm;
        fin >> closed_farm;
        closed_farm--; // Adjusting to 0-based indexing
        int farm_region_id = farm_to_region_id[closed_farm];
        vector<int> farms_in_same_region = region_id_to_farm[farm_region_id];
        closed[closed_farm] = true;
        visited = closed;

        region_id_to_farm.erase(farm_region_id);

        for (int farm : farms_in_same_region) {
            if (!visited[farm]) {
                region_id_to_farm[region_id] = vector<int>();
                dfs(farm, region_id);
                region_id++;
            }
        }

        if (region_id_to_farm.size() == 1) {
            fout << "YES" << endl;
        } else {
            fout << "NO" << endl;
        }
    }

    fout.close();

    return 0;
}

void dfs(int current, int region_id) {
    if (visited[current]) {
        return;
    }

    visited[current] = true;
    farm_to_region_id[current] = region_id;
    region_id_to_farm[region_id].push_back(current);

    for (int next : graph[current]) {
        if (!visited[next]) {
            dfs(next, region_id);
        }
    }
}
